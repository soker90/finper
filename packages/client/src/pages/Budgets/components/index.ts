export { default as BudgetTable } from './BudgetTable'
export { default as BudgetCard } from './BudgetCard'
