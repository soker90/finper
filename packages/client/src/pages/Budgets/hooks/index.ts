export { useBudgets } from './useBudgets'
export { useBudgetDataCard } from './useDataCard'
