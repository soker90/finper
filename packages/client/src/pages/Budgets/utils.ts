export const isSameDate = (year?: string, month?: string): boolean => {
  const now = new Date()
  return (year === now.getFullYear().toString() && month === now.getMonth().toString())
}

export const getUrlMonth = (year?: string, month?: number) => `/presupuestos/${year}/${month}`
