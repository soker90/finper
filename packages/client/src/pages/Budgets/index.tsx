import { useMemo } from 'react'
import { useParams } from 'react-router-dom'
import { Grid } from '@mui/material'
import { useBudgets } from './hooks'

import { BudgetCard, BudgetTable } from './components'
import Header from './components/Header'

const Budgets = () => {
  const { year, month } = useParams()
  const { expenses, incomes } = useBudgets({ year, month })

  const expensesTotal = useMemo(() => {
    const total = expenses?.reduce((sum, { budgets }) => sum + budgets?.[0]?.real, 0) ?? 0
    const estimated = expenses?.reduce((sum, { budgets }) => sum + budgets?.[0]?.amount, 0) ?? 0
    return {
      total,
      estimated,
      percentage: (total / estimated) * 100,
      isPositive: total <= estimated
    }
  }, [expenses])

  const incomesTotal = useMemo(() => {
    const total = incomes?.reduce((sum, { budgets }) => sum + budgets?.[0]?.real, 0) ?? 0
    const estimated = incomes?.reduce((sum, { budgets }) => sum + budgets?.[0]?.amount, 0) ?? 0
    return {
      total,
      estimated,
      percentage: (total / estimated) * 100,
      isPositive: total <= estimated
    }
  }, [incomes])

  const balanceTotal = incomesTotal.total - expensesTotal.total
  const balanceEstimated = incomesTotal.estimated - expensesTotal.estimated
  const balancePercentage = (balanceTotal / balanceEstimated) * 100

  return (
        <>
            <Header month={month} year={year}/>
            <Grid container spacing={3}>
                <Grid item xs={12} md={4}>
                    <BudgetCard title='Gastos' total={expensesTotal.total} estimated={expensesTotal.estimated}
                                percentage={expensesTotal.percentage} color='warning'
                                isPositive={expensesTotal.isPositive}/>
                </Grid>
                <Grid item xs={12} md={4}>
                    <BudgetCard title='Ingresos' total={incomesTotal.total} estimated={incomesTotal.estimated}
                                percentage={incomesTotal.percentage} color='success'
                                isPositive={incomesTotal.isPositive}/>
                </Grid>
                <Grid item xs={12} md={4}>
                    <BudgetCard title='Balance' total={incomesTotal.total - expensesTotal.total}
                                estimated={incomesTotal.estimated - expensesTotal.estimated}
                                percentage={balancePercentage} isPositive={balancePercentage <= 100}/>
                </Grid>

                {year &&
                  <>
                    <BudgetTable budgets={expenses} title={'Gastos'} year={year} month={month as string}/>
                    <BudgetTable budgets={incomes} title={'Ingresos'} year={year} month={month as string}/>
                  </>
                }
            </Grid>
        </>
  )
}

export default Budgets
