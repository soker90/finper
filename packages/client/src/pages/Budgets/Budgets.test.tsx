// @vitest-environment happy-dom
import { beforeEach, describe, expect, it, vi } from 'vitest'

import { render } from '../../test/testUtils'

import Budgets from './index'

describe('Budgets', async () => {
  beforeEach(() => {
    vi.mock('react-router', async () => {
      const reactRouterDom = await vi.importActual('react-router')
      return ({
        ...reactRouterDom as any,
        useParams: vi.fn(() => ({
          year: '2022',
          month: '8'
        }))
      })
    })
  })
  it('Show title success', async () => {
    const { findByText } = render(<Budgets />)
    const monthLabel = await findByText('Septiembre 2022')

    expect(monthLabel).toBeDefined()
  })
})
