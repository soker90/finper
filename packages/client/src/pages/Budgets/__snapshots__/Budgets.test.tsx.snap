// Vitest Snapshot v1

exports[`Budgets > Snapshot empty case 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-4rifo3-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <a
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation css-oqu2pi-MuiButtonBase-root-MuiButton-root"
        href="/presupuestos/2022/9"
        tabindex="0"
      >
         ACTUAL
      </a>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        class="css-1p1m4ay"
      >
        <a
          aria-label="izquierda"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeLarge css-172ns19-MuiButtonBase-root-MuiIconButton-root"
          href="/presupuestos/2022/7"
          tabindex="0"
        >
          <span
            aria-label="left"
            class="anticon anticon-left"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="left"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
              />
            </svg>
          </span>
        </a>
        <p
          class="MuiTypography-root MuiTypography-body1 css-oub1po-MuiTypography-root"
        >
          Septiembre
           
          2022
           
        </p>
        <a
          aria-label="derecha"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeLarge css-172ns19-MuiButtonBase-root-MuiIconButton-root"
          data-testid="right-arrow"
          href="/presupuestos/2022/9"
          tabindex="0"
        >
          <span
            aria-label="right"
            class="anticon anticon-right"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="right"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
              />
            </svg>
          </span>
        </a>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation css-oqu2pi-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Copiar mes anterior 
      </button>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-1lwnmxi-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-13mu4hp-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-1uvcgxq-MuiCardContent-root"
        >
          <div
            class="css-zgbp5c-MuiStack-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-8qynzz-MuiTypography-root"
            >
              Gastos
            </h6>
            <div
              class="MuiGrid-root MuiGrid-container css-1vam7s3-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <h4
                  class="MuiTypography-root MuiTypography-h4 css-10zzhnw-MuiTypography-root"
                  data-testid="total-expenses"
                >
                  0 €
                </h4>
              </div>
            </div>
          </div>
          <div
            class="MuiBox-root css-1bihoop"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-1lifr75-MuiTypography-root"
            >
              Estimado
               
              <span
                class="MuiTypography-root MuiTypography-caption css-hvj5iu-MuiTypography-root"
              >
                0 €
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-1lwnmxi-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-13mu4hp-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-1uvcgxq-MuiCardContent-root"
        >
          <div
            class="css-zgbp5c-MuiStack-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-8qynzz-MuiTypography-root"
            >
              Ingresos
            </h6>
            <div
              class="MuiGrid-root MuiGrid-container css-1vam7s3-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <h4
                  class="MuiTypography-root MuiTypography-h4 css-10zzhnw-MuiTypography-root"
                  data-testid="total-incomes"
                >
                  0 €
                </h4>
              </div>
            </div>
          </div>
          <div
            class="MuiBox-root css-1bihoop"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-1lifr75-MuiTypography-root"
            >
              Estimado
               
              <span
                class="MuiTypography-root MuiTypography-caption css-1wfx628-MuiTypography-root"
              >
                0 €
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-1lwnmxi-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-13mu4hp-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-1uvcgxq-MuiCardContent-root"
        >
          <div
            class="css-zgbp5c-MuiStack-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-8qynzz-MuiTypography-root"
            >
              Balance
            </h6>
            <div
              class="MuiGrid-root MuiGrid-container css-1vam7s3-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <h4
                  class="MuiTypography-root MuiTypography-h4 css-10zzhnw-MuiTypography-root"
                  data-testid="total-total"
                >
                  0 €
                </h4>
              </div>
            </div>
          </div>
          <div
            class="MuiBox-root css-1bihoop"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-1lifr75-MuiTypography-root"
            >
              Estimado
               
              <span
                class="MuiTypography-root MuiTypography-caption css-2hztrc-MuiTypography-root"
              >
                0 €
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-6 css-va0ja1-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1908txl-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiBox-root css-19midj6"
        >
          <div
            class="MuiBox-root css-70qvj9"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-12orxl-MuiTypography-root"
            >
              Gastos
            </h4>
          </div>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <table
            class="MuiTable-root css-d7ymjm-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-110l8om-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1j77n0n-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-tvar7v-MuiTableCell-root"
                  scope="col"
                >
                  Categoría
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-tvar7v-MuiTableCell-root"
                  scope="col"
                >
                  Real
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-tvar7v-MuiTableCell-root"
                  scope="col"
                >
                  Estimado
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1l6v1u7-MuiTableCell-root"
                  scope="col"
                >
                  Acciones
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-1geald0-MuiTableBody-root"
            />
          </table>
          <div
            class="MuiBox-root css-19midj6"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter css-1br5zzv-MuiTypography-root"
            >
              No se han encontrado datos
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-6 css-va0ja1-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1908txl-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiBox-root css-19midj6"
        >
          <div
            class="MuiBox-root css-70qvj9"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-12orxl-MuiTypography-root"
            >
              Ingresos
            </h4>
          </div>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <table
            class="MuiTable-root css-d7ymjm-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-110l8om-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1j77n0n-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-tvar7v-MuiTableCell-root"
                  scope="col"
                >
                  Categoría
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-tvar7v-MuiTableCell-root"
                  scope="col"
                >
                  Real
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-tvar7v-MuiTableCell-root"
                  scope="col"
                >
                  Estimado
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1l6v1u7-MuiTableCell-root"
                  scope="col"
                >
                  Acciones
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-1geald0-MuiTableBody-root"
            />
          </table>
          <div
            class="MuiBox-root css-19midj6"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter css-1br5zzv-MuiTypography-root"
            >
              No se han encontrado datos
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
